Cython/Compiler/ExprNodes.py

```diff
@@ -14131,20 +14131,29 @@ def analyse_types(self, env):
 
 
 def coerce_from_soft_complex(arg, dst_type, env):
+    nogil = env.nogil or getattr(arg, "nogil", False)
```

Comment:
Using `getattr()` in this way always has a code smell. It suggests that we are relying on or introducing special cases that are not well defined. In the call to `binop_node()`, you pass `node.nogil` unconditionally, whereas here, you are not asserting that the attribute exists.

I can see that some `ExprNode`s already have such an attribute. It feels like a slightly clearer name might have been better (Is this node used inside of a nogil section, or does it start one? *Can* it be executed without the GIL or *must* it run without it?), but that's what we have now. If we need to extend what we have, then we should try to stick to the semantics or make them clearer. Maybe `nogil` is a general attribute of an expression node? And maybe we can document it right in `ExprNode` then? Maybe that even allows us to add a general nogil check at that level that subclasses only need to override at need? (The latter might not be for 3.0 any more…)

Reply:
I think:

Coerce can be at a number of different times in the pipeline. `env.nogil` is only reliable in some transforms. `node.nogil` is a general attribute of an expression node, but is set fairly late on pipeline so isn't always available. So I was trying to work around that mess.

With that said, I think I have a better way to do it in this case (it can just be deferred until code generation time, when it's definitely known) since we don't make any decisions based on it except writing 0 or 1

Reply:
I think the best long-term solution is to add `nogil` as a general attribute of a (Expr?)Node.  Or possibly just try to get it from `code` at code-generation time in a few more places.  For now I've added something that I think is slightly less less bad than I had before (It at least has the advantage that all the hackyness is concentrated in one place, which makes it easier to reason about)